### PR TITLE
Redefine operations

### PR DESCRIPTION
**Pull request by lintusj1**
Allows one to write `t1.redefine(operation_fn, parents)`. Also lots of refactoring. Now the name _transform_ refers to callables that turn `input_dict` to `output_dict` and the name _operation_ referes to callables that produce `data` for the `output_dict` (e.g. simulator functions, summary functions etc.)

### CODE REVIEW

File: elfi/distributions.py [hunk 1]
```diff
@@ -7,32 +7,34 @@
...
-    size = (input_dict['n'],)+tuple(size)
-    data = distribution.rvs(*input_dict['data'], size=size, random_state=prng)
+    batch_size = input_dict["n"]
+    data = operation(*input_dict['data'], batch_size=batch_size, random_state=prng)
     return core.to_output_dict(input_dict, data=data, random_state=prng.get_state())
 
 
-class ScipyRV(core.RandomStateMixin, core.Operation):
+class Distribution:
+    """Must have an attribute or property `name`
```
Comment (akangasr) [inline elfi/distributions.py]: Docstring should explain the purpose of the class as well (eg. do we expect that all distributions inherit this interface?) Also as this is an interface class all the methods should have proper docstrings as well to explain expected behavior.

File: elfi/tools.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+from functools import partial
```
Comment (akangasr) [inline elfi/tools.py]: Don't we already have utils.py? What is the difference of utils and tools?

File: elfi/tools.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import numpy as np
+
+# TODO: make a cleaner implementation
+def vectorized_operation(operation, *input_data, batch_size=None, **kwargs):
```
Comment (akangasr) [inline elfi/tools.py]: Docstring missing. Also, should this now be vectorized_transform?

File: notebooks/ma2_example.ipynb [hunk 4]
```diff
@@ -900,21 +910,21 @@
  "metadata": {
   "anaconda-cloud": {},
   "kernelspec": {
-   "display_name": "Python 3",
+   "display_name": "Python [conda env:abc4py]",
```
Comment (akangasr) [inline notebooks/ma2_example.ipynb]: These metadata should not be in the changeset.

File: tests/unit/test_decorators.py [hunk 5]
```diff
@@ -1,58 +0,0 @@
-import numpy as np
```
Comment (akangasr) [inline tests/unit/test_decorators.py]: These should not be removed but changed to follow the new implementation (eg. testing that the tools.vectorize works with all relevant node types).

File: elfi/core.py [hunk 6]
```diff
@@ -256,8 +256,8 @@ def normalize_data_dict(dict, n):
     return normalized
 
 
-class Operation(Node):
-    def __init__(self, name, operation, *parents, inference_task=None, store=None):
+class Transform(Node):
+    def __init__(self, name, transform, *parents, inference_task=None, store=None):
         """Operation node transforms data from parents to output
```
Comment (lintusj1) [inline elfi/core.py]: Done.

File: elfi/core.py [hunk 7]
```diff
@@ -387,10 +384,20 @@ def inference_task(self):
...
     @property
+    def transform(self):
+        return self._transform
+
+    @property
     def version(self):
         """Version of the node (currently number of resets)"""
         return self._num_resets
 
+    def redefine(self, transform, *parents):
```
Comment (lintusj1) [inline elfi/core.py]: Done.

File: elfi/core.py [hunk 8]
```diff
@@ -440,13 +447,40 @@ def _create_delayed_output(self, sl, input_dict, with_values=None):
...
-            return delayed(self.operation)(dinput,
-                                           dask_key_name=dask_key)
+            return delayed(self._transform)(dinput,
+                                            dask_key_name=dask_key)
 
     def _convert_to_node(self, obj, name):
         return Constant(name, obj)
 
 
+class Operation(Transform):
```
Comment (lintusj1) [inline elfi/core.py]: Done.

File: elfi/core.py [hunk 9]
```diff
@@ -511,7 +545,7 @@ def _inherit_observed(self):
             if not hasattr(parent, "observed"):
                 raise ValueError("Parent {} has no observed value to inherit".format(parent))
         observed = tuple([p.observed for p in self.parents])
-        observed = self.operation({"data": observed, "n": 1})["data"]
+        observed = self._transform({"data": observed, "n": 1})["data"]
```
Comment (lintusj1) [inline elfi/core.py]: Maybe, currently I don't want it to be changed without the core knowing about it.

File: elfi/distributions.py [hunk 10]
```diff
@@ -7,32 +7,34 @@
 from . import core
 
 
-def spr_op(distribution, size, input_dict):
+# TODO: combine with `simulator_wrapper` and perhaps with `summary_wrapper`?
+def random_wrapper(operation, input_dict):
```
Comment (lintusj1) [inline elfi/distributions.py]: Added

File: elfi/distributions.py (hunk 1 above)
Comment (lintusj1) [inline elfi/distributions.py]: Done.

File: elfi/tools.py (hunk 2 above)
Comment (lintusj1) [inline elfi/tools.py]: Utils would be for internal use, tools for those that use elfi.

File: notebooks/ma2_example.ipynb (hunk 4 above)
Comment (lintusj1) [inline notebooks/ma2_example.ipynb]: Changed back.

File: tests/unit/test_decorators.py (hunk 5 above)
Comment (lintusj1) [inline tests/unit/test_decorators.py]: Done.

**Pull request merged**